File: backend/src/services/realtime.service.ts
import { IoTDataPlaneClient, PublishCommand } from '@aws-sdk/client-iot-data';
import { config } from '../config';
import WebSocket from 'ws';

// AWS IoT Core client for real-time messaging
const iotClient = new IoTDataPlaneClient({
  region: config.aws.region,
  credentials: {
    accessKeyId: config.aws.accessKeyId!,
    secretAccessKey: config.aws.secretAccessKey!,
  },
  endpoint: `https://${config.iot.endpoint}`,
});

export class RealtimeService {
  private wsServer: WebSocket.Server | null = null;

  // Initialize WebSocket server for client connections
  initializeWebSocketServer(server: any) {
    this.wsServer = new WebSocket.Server({ server });

    this.wsServer.on('connection', (ws: WebSocket, req) => {
      console.log('Client connected to WebSocket');

      ws.on('message', (message: string) => {
        console.log('Received:', message);
      });

      ws.on('close', () => {
        console.log('Client disconnected');
      });
    });
  }

  // Broadcast report changes to all connected clients via AWS IoT
  async broadcastReportChange(reportId: string, change: any) {
    const topic = `reports/${reportId}/updates`;
    const payload = JSON.stringify({
      reportId,
      timestamp: new Date().toISOString(),
      change,
    });

    try {
      await iotClient.send(
        new PublishCommand({
          topic,
          payload: Buffer.from(payload),
          qos: 1,
        })
      );

      // Also broadcast to local WebSocket clients
      this.broadcastToLocalClients(reportId, change);

      console.log(`Broadcasted change to topic: ${topic}`);
    } catch (error) {
      console.error('Failed to broadcast report change:', error);
      throw error;
    }
  }

  // Broadcast to local WebSocket clients
  private broadcastToLocalClients(reportId: string, change: any) {
    if (!this.wsServer) return;

    const message = JSON.stringify({
      type: 'report_update',
      reportId,
      change,
      timestamp: new Date().toISOString(),
    });

    this.wsServer.clients.forEach((client) => {
      if (client.readyState === WebSocket.OPEN) {
        client.send(message);
      }
    });
  }

  // Subscribe to report updates (for server-side listeners)
  async subscribeToReport(reportId: string, callback: (change: any) => void) {
    // In production, this would use AWS IoT MQTT subscriptions
    // For now, we'll use local WebSocket connections
    console.log(`Subscribed to report: ${reportId}`);
  }
}

export const realtimeService = new RealtimeService();

File: backend/app/services/realtime_service.py
import boto3
import json
from datetime import datetime
from botocore.exceptions import ClientError
from app.config import get_settings

settings = get_settings()

# Initialize IoT Data client
iot_client = boto3.client(
    'iot-data',
    region_name=settings.aws_region,
    aws_access_key_id=settings.aws_access_key_id,
    aws_secret_access_key=settings.aws_secret_access_key,
    aws_session_token=settings.aws_session_token,
    endpoint_url=f"https://{settings.aws_iot_endpoint}"
)


class RealtimeService:
    @staticmethod
    async def broadcast_report_change(report_id: str, change: dict) -> None:
        """Broadcast report changes to all connected clients via AWS IoT"""
        topic = f"reports/{report_id}/updates"
        payload = {
            "report_id": report_id,
            "timestamp": datetime.now().isoformat(),
            "change": change,
        }
        
        try:
            iot_client.publish(
                topic=topic,
                qos=1,
                payload=json.dumps(payload)
            )
            print(f"Broadcasted change to topic: {topic}")
        except ClientError as e:
            print(f"Failed to broadcast report change: {e}")
            raise
    
    @staticmethod
    async def send_notification(org_id: str, notification: dict) -> None:
        """Send organization-wide notification"""
        topic = f"organizations/{org_id}/notifications"
        payload = {
            "org_id": org_id,
            "timestamp": datetime.now().isoformat(),
            "notification": notification,
        }
        
        try:
            iot_client.publish(
                topic=topic,
                qos=1,
                payload=json.dumps(payload)
            )
            print(f"Sent notification to topic: {topic}")
        except ClientError as e:
            print(f"Failed to send notification: {e}")
            raise
    
    @staticmethod
    async def send_user_message(user_id: str, message: dict) -> None:
        """Send message to specific user"""
        topic = f"users/{user_id}/messages"
        payload = {
            "user_id": user_id,
            "timestamp": datetime.now().isoformat(),
            "message": message,
        }
        
        try:
            iot_client.publish(
                topic=topic,
                qos=1,
                payload=json.dumps(payload)
            )
            print(f"Sent message to topic: {topic}")
        except ClientError as e:
            print(f"Failed to send user message: {e}")
            raise
